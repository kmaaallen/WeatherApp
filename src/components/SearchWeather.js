import React, { useEffect, useLayoutEffect, useState, useRef } from 'react';
//Material UI
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';

const useStyles = makeStyles({
    form: {
        padding: '15px',
        textAlign: 'right',
        width: '50%'
    },
    search: {
        minWidth: '300px',
        background: '#fff',
        borderRadius: '4px'
    },
    button: {
        marginLeft: '15px',
        marginTop: '6px',
        height: '40px'
    },
    blackText: {
        color: 'black'
    }
})

function Search(props) {
    const classes = useStyles();
    const [city, setCity] = useState('');
    const [cityData, setCityData] = useState([]);
    const [searchError, setSearchError] = useState('');

    function handleSubmit(e) {
        setSearchError('');
        e.preventDefault();
        fetch(`/api/weather/${city}`)
            .then((res) => res.json())
            .then((cityData) => {
                if (cityData.data.cod === "404") {
                    var capitalised = cityData.data.message.charAt(0).toUpperCase() + cityData.data.message.slice(1);
                    setSearchError(capitalised);
                } else {
                    setCityData(cityData.data);
                    setCity('');
                }
            })
            .catch(error => {
                console.error(error);
            })
    }

    const prevCityData = useRef();
    useEffect(() => {
        prevCityData.current = cityData;
    });

    const prevCity = prevCityData.current;

    useLayoutEffect(() => {
        if (cityData === prevCity) {
            return;
        } else {
            props.callBackFromParent(cityData);
        }
    }, [props, cityData, prevCity])

    return (
        <form className={classes.form} onSubmit={handleSubmit}>
            <TextField
                id="city-text"
                onChange={(event) => { setCity(event.target.value) }}
                label="City"
                InputLabelProps={{ className: classes.blackText }} // find prop for this
                value={city}
                error={searchError ? true : false}
                helperText={searchError}
                variant="filled"
                className={classes.search} />
            <Button
                type="submit"
                variant="contained"
                className={classes.button}
            >Search</Button>
        </form>
    )
}

export default Search;