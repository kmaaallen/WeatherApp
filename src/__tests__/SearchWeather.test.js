import React from 'react';
import renderer from 'react-test-renderer';
import SearchWeather from '../components/SearchWeather';

// mock functions
const myCallBack = jest.fn();

describe('<SearchWeather />', () => {
    it('renders correctly', () => {
        const tree = renderer.create(<SearchWeather callBackFromParent={myCallBack} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
