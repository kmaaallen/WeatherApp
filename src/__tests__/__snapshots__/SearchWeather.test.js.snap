// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<form
  className="makeStyles-form-1"
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-search-2"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
      data-shrink={false}
      htmlFor="city-text"
      id="city-text-label"
    >
      City
    </label>
    <div
      className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiFilledInput-input"
        disabled={false}
        id="city-text"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
    </div>
    
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-3"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiButton-label"
    >
      Search
    </span>
  </button>
</form>
`;

exports[`<SearchWeather /> renders correctly 1`] = `
<form
  className="makeStyles-form-1"
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-search-2"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
      data-shrink={false}
      htmlFor="city-text"
      id="city-text-label"
    >
      City
    </label>
    <div
      className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiFilledInput-input"
        disabled={false}
        id="city-text"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
    </div>
    
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-3"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiButton-label"
    >
      Search
    </span>
  </button>
</form>
`;
